src/shims/unix/socket.rs

```diff
@@ -39,15 +54,71 @@ impl FileDescription for SocketPair {
         "socketpair"
     }
 
+    fn return_epoll_events<'tcx>(&mut self) -> InterpResult<'tcx, &mut Vec<Weak<EpollEvent>>> {
+        Ok(&mut self.epoll_events)
+    }
+
+    fn check_readiness<'tcx>(&self, ecx: &mut MiriInterpCx<'tcx>) -> InterpResult<'tcx, Vec<u32>> {
+        // We only check the status of EPOLLIN, EPOLLOUT and EPOLLRDHUP flag. If other event flags
+        // need to be supported in the future, the check should be added here.
+        let epollin = ecx.eval_libc_u32("EPOLLIN");
+        let epollout = ecx.eval_libc_u32("EPOLLOUT");
+        let epollrdhup = ecx.eval_libc_u32("EPOLLRDHUP");
+        let mut ready_flags = Vec::new();
+        let readbuf = self.readbuf.borrow();
+        // Check if the writeend has closed.
+        if !readbuf.buf_has_writer {
+            ready_flags.push(epollrdhup);
+        }
+
+        // Check if it is readable.
+        if !readbuf.buf.is_empty() {
+            ready_flags.push(epollin);
+        }
+
+        // Check if is writable.
+        if let Some(writebuf) = self.writebuf.upgrade() {
+            let writebuf = writebuf.borrow();
+            let data_size = writebuf.buf.len();
+            let available_space = MAX_SOCKETPAIR_BUFFER_CAPACITY.strict_sub(data_size);
+            if available_space != 0 {
+                ready_flags.push(epollout);
+            }
+        }
+
+        // Check if the peer_fd closed
+        if self.peer_closed {
+            ready_flags.push(epollrdhup);
+            // This is an edge case. Whenever epollrdhup is triggered, epollin will be added
+            // even though there is no data in the buffer.
+            ready_flags.push(epollin);
+        }
+
+        Ok(ready_flags)
+    }
+
     fn close<'tcx>(
         self: Box<Self>,
+        ecx: &mut MiriInterpCx<'tcx>,
         _communicate_allowed: bool,
     ) -> InterpResult<'tcx, io::Result<()>> {
         // This is used to signal socketfd of other side that there is no writer to its readbuf.
         // If the upgrade fails, there is no need to update as all read ends have been dropped.
         if let Some(writebuf) = self.writebuf.upgrade() {
             writebuf.borrow_mut().buf_has_writer = false;
         };
+
+        // Notify peer fd that closed has happened.
+        // This check is only for linux because other target have no libc epoll flags.
+        if ecx.tcx.sess.target.os == "linux" {
```

Comment:
In this case we would have an empty `epoll_events`, so while we'd do some work to compute the ready flags, we'd perform no action at all

Reply:
Yup I expect this to happen for all non-linux target. I haven't checked which non-linux target can use ``eventfd`` and ``socketpair``, but I wish we can still support ``socketpair`` and ``eventfd`` on these targets even if these targets don't use ``epoll``.